crates/red_knot/src/semantic/types.rs

```diff
@@ -27,6 +27,8 @@ pub enum Type {
     Unknown,
     /// name is not bound to any value
     Unbound,
+    /// the None object
+    None,
```

Comment:
I'm okay with this if it unblocks your work on type narrowing, but in the long term, I'm not sure we should represent `None` as a distinct type. Mypy and pyright have both had a lot of bugs because of the extent to which they special-case `None` (pyright particularly), and I'd like us to think hard about whether we need to do so to the same extent that they've done.

What's actually special about `None`? Three things, I think:
1. When we see `None` in a type annotation, we should immediately desugar that to `Instance(types.NoneType)` (or [the typeshed backport](https://github.com/python/typeshed/blob/4da35728277ee15c16f578ecf62acfc71b8caadf/stdlib/_typeshed/__init__.pyi#L308-L311) for Python less than 3.10)
2. When printing the type to users in diagnostics, we should immediately resugar it from `Instance(types.NoneType)` to `None`
3. We need to understand internally that `NoneType` has exactly one instance, `None`, so that you can use identity checks such as `x is not None` for type narrowing

Well, I say "three things", but actually, the third point isn't unique at all! Other singletons exist in Python, such as `...`, `NotImplemented`, `typing.NoDefault` on py313+, etc. Mypy and pyright both support narrowing unions including `types.EllipsisType` with `x is not ...`, and I think we should too. And maybe it would be reasonable for a user to narrow a `bool | str` union via a `x is not True and x is not False` check.

So does `None` really need to be its own variant in the enum here? Or should we have a more generalised way of representing builtin types that only have 1 or 2 possible instances, and that can therefore be narrowed via identity checks?

As I say, I'm fine with this for now, though.

Reply:
This is a good question, and I debated the same thing while working on this PR.

I went with this for now basically because otherwise I would have had to wait for typeshed to be ready, and I needed some representation of `None` for type narrowing to work. But I think it will probably be better to switch to `Instance(types.NoneType)` once we have typeshed in place. I'll add a TODO for that.

Reply:
Great, sounds like we're aligned. Thanks!

Reply:
I'll have typeshed for you soon!!